framework/crud_resource.go

```diff
@@ -0,0 +1,406 @@
+	// already done at this point of the reconciliation loop.
+	ApplyDeleteChange(ctx context.Context, obj, deleteChange interface{}) error
+	// ApplyUpdateChange receives the new custom object observed during
+	// custom resource watches. It also receives the update portion of the
+	// Patch provided by NewUpdatePatch or NewDeletePatch.
+	// ApplyUpdateChange has to update resources based on its provided
+	// input. All other reconciliation logic and state transformation is
+	// already done at this point of the reconciliation loop.
+	ApplyUpdateChange(ctx context.Context, obj, updateChange interface{}) error
+}
+
+type CRUDResource struct {
+	ops CRUDResourceOps
+}
+
+func NewCRUDResource(config CRUDResourceConfig) (*CRUDResource, error) {
+	if config.Ops == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Ops must not be empty")
+	}
+	if config.Ops.Name() == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Ops.Name() must not be empty")
+	}
+	if config.Ops.Underlying() == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Ops.Underlying() must not be empty")
+	}
+
+	r := &CRUDResource{
+		ops: config.Ops,
+	}
+
+	return r, nil
+}
+
+func (r *CRUDResource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	var err error
+
+	var currentState interface{}
+	{
+		if reconciliationcanceledcontext.IsCanceled(ctx) {
+			return nil
+		}
+		if resourcecanceledcontext.IsCanceled(ctx) {
+			ctx = resourcecanceledcontext.NewContext(ctx, make(chan struct{}))
+			return nil
+		}
+
+		meta, ok := loggermeta.FromContext(ctx)
+		if ok {
+			meta.KeyVals["function"] = "GetCurrentState"
+			defer delete(meta.KeyVals, "function")
+		}
+		currentState, err = r.ops.GetCurrentState(ctx, obj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var desiredState interface{}
+	{
+		if reconciliationcanceledcontext.IsCanceled(ctx) {
+			return nil
+		}
+		if resourcecanceledcontext.IsCanceled(ctx) {
+			ctx = resourcecanceledcontext.NewContext(ctx, make(chan struct{}))
+			return nil
+		}
+
+		meta, ok := loggermeta.FromContext(ctx)
+		if ok {
+			meta.KeyVals["function"] = "GetDesiredState"
+			defer delete(meta.KeyVals, "function")
+		}
+		desiredState, err = r.ops.GetDesiredState(ctx, obj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var patch *Patch
+	{
+		if reconciliationcanceledcontext.IsCanceled(ctx) {
+			return nil
+		}
+		if resourcecanceledcontext.IsCanceled(ctx) {
+			ctx = resourcecanceledcontext.NewContext(ctx, make(chan struct{}))
+			return nil
+		}
+
+		meta, ok := loggermeta.FromContext(ctx)
+		if ok {
+			meta.KeyVals["function"] = "NewUpdatePatch"
+			defer delete(meta.KeyVals, "function")
+		}
+		patch, err = r.ops.NewUpdatePatch(ctx, obj, currentState, desiredState)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if patch == nil {
+			return microerror.Maskf(executionFailedError, "patch must not be nil")
+		}
```

Comment:
This here changed in master now. Lets please align it to allow a `nil` return value for empty patches.

Reply:
Done.